### PR TITLE
Implementation of TCP Sockets and InetSocketAddress

### PR DESCRIPTION
**Pull request by asoltysik**
Ready to review. The PR contains a lot of commits from the InetAddress PR. InetSocketAddress ported from Harmony, some of the code for Socket also inspired by Harmony. There are some stuff missing and commented out, which can be filled in the next PR, mainly: 

- `java.net.Socket` supports only streaming mode, throws `UnsupportedOperationException` otherwise. This can be implemented together with Datagram Sockets.
- One socket option: `SO_BINDADDR`, as I have trouble understanding its purpose. Also this option cannot be set or gotten.
- No OOB data sending for now.
- No sockets through `java.net.Proxy` (therefore also no SOCKS protocol support).

### CODE REVIEW

**Comment by asoltysik** [comment]
This is ready to review right after `InetAddress` PR is dealt with.

**Comment by densh** [comment]
@asoltysik Can you please squash your changes and rebase on top of latest master, thanks!

**Comment by asoltysik** [comment]
Looks like CI fails on Macs only and it fails because of either `connect` or `bind` with timeout. It may have something to do with `select` binding. I don't own a Mac so I can't test it manually. I will try one other thing, but if somebody could check for a possible cause on macs that would be great.

**Comment by densh** [comment]
@asoltysik I just restarted the timed out builds. 👍

File: javalib/src/main/scala/java/net/SocketInputStream.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package java.net
+
+import java.io.InputStream
+
+// Ported from Apache Harmony
+private[net] class SocketInputStream(socket: PlainSocketImpl)
+    extends InputStream {
+
+  override def close: Unit = socket.close
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/SocketInputStream.scala]: `close()`

File: javalib/src/main/scala/java/net/SocketOutputStream.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package java.net
+
+import java.io.OutputStream
+
+// Ported from Apache Harmony
+private[net] class SocketOutputStream(socket: PlainSocketImpl)
+    extends OutputStream {
+
+  override def close: Unit = {
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/SocketOutputStream.scala]: `close()`

File: scripted-tests/run/java-net-socket/build.sbt [hunk 3]
```diff
@@ -9,10 +9,40 @@ enablePlugins(ScalaNativePlugin)
 
 scalaVersion := "2.11.11"
 
-lazy val launchEchoServer =
-  taskKey[Unit]("Setting up tcp echo server")
+lazy val launchServer = taskKey[Unit]("Setting up a server for tests")
+lazy val launchTcpEchoServer =
+  taskKey[Unit]("Setting up echo server on port 5832")
```
Comment (Duhemm) [inline scripted-tests/run/java-net-socket/build.sbt]: The doc is outdated (the port is not fixed)

File: test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala [hunk 4]
```diff
@@ -7,19 +7,14 @@ import java.io.{
   DataInputStream,
   DataOutputStream
 }
+import java.net.Socket
```
Comment (Duhemm) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: Yay! 👍

File: test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala [hunk 5]
```diff
@@ -117,55 +80,46 @@ abstract class TestMainBase {
...
-    val msglen = read(client)(4).readInt()
-    val in     = read(client)(msglen)
+  private def receive[T](dis: DataInputStream): Message = {
+    val msglen = dis.readInt()
     val msgbuf = new Array[Byte](msglen)
-    var total  = 0
-    while (total < msglen) {
-      total += in.read(msgbuf, total, msglen - total)
+    if (msglen > 0) {
+      dis.readFully(msgbuf)
```
Comment (Duhemm) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: I'm afraid you may read 2 messages at once here, but only the first will get parsed, and the second will be lost. It's better to keep using `msglen` to know where to stop.

Comment (asoltysik) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: The number of bytes read by `dis.readFully` is the length of the buffer, which in this case is `msglen`: 
``` scala
val msgbuf = new Array[Byte](msglen)
```

Comment (asoltysik) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: https://docs.oracle.com/javase/8/docs/api/java/io/DataInputStream.html#readFully-byte:A-

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+import scala.scalanative.posix.netdb._
+import scala.scalanative.posix.netdbOps._
+import scala.scalanative.posix.sys.ioctl._
+import scala.scalanative.posix.fcntl._
+import scala.scalanative.posix.sys.select._
+import scala.scalanative.posix.sys.selectOps._
+import scala.scalanative.posix.unistd.{close => cClose}
+import java.io.{FileDescriptor, IOException, OutputStream, InputStream}
+
+private[net] class PlainSocketImpl extends SocketImpl {
```
Comment (asoltysik) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: Not really, I may have been inspired by Harmony but the implementation is different in most cases. Should I mention it here?

File: test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala (hunk 5 above)
Comment (Duhemm) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: Oh yeah sure you're right!

**Comment by densh** [comment]
Please rebase your PR on top of latest master due to CI changes, thanks!

**Comment by asoltysik** [comment]
@densh done

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala (hunk 6 above)
Comment (densh) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: If it's not a 1-to-1 port, you don't have to.

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+      }
+      sent
+    }
+  }
+
+  def read(buffer: Array[Byte], offset: Int, count: Int): Int = {
+    if (shutInput) -1
+
+    val cBuff    = stackalloc[Byte](count)
+    val bytesNum = socket.recv(fd.fd, cBuff, count, 0).toInt
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: Same here. You can directly do
```scala
val bytesNum = socket.recv(fd.fd, buffer.asInstanceOf[ByteArray].at(offset), count, 0).toInt
```

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+    case SocketOptions.IP_TOS       => in.IP_TOS
+    case SocketOptions.SO_KEEPALIVE => socket.SO_KEEPALIVE
+    case SocketOptions.SO_LINGER    => socket.SO_LINGER
+    case SocketOptions.SO_TIMEOUT   => socket.SO_RCVTIMEO
+    case SocketOptions.SO_OOBINLINE => socket.SO_OOBINLINE
+    case SocketOptions.SO_RCVBUF    => socket.SO_RCVBUF
+    case SocketOptions.SO_SNDBUF    => socket.SO_SNDBUF
+    case SocketOptions.SO_REUSEADDR => socket.SO_REUSEADDR
+    case SocketOptions.TCP_NODELAY  => tcp.TCP_NODELAY
+    case _                          => throw new Error("This shouldn't happen")
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]:
```scala
case _ => sys.error(s"Unknown option: $option)
```

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+      if (linger.l_onoff != 0) {
+        Integer.valueOf(linger.l_linger)
+      } else {
+        Integer.valueOf(-1)
+      }
+    } else if (optID == SocketOptions.SO_TIMEOUT) {
+      Integer.valueOf(this.timeout)
+    } else {
+      Integer.valueOf(!(opt.cast[Ptr[CInt]]))
+    }
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: The whole expression is a bit hard to follow. Could you do:

```scala
optID match {
  case SocketOptions.TCP_NODELAY | SocketOptions.SO_KEEPALIVE | ... =>
     // ...
   case SocketOptions.SO_LINGER =>
     // ...
   case ...
}
```

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+
+      ptr.tv_sec = mseconds / 1000
+      ptr.tv_usec = (mseconds % 1000) * 1000
+
+      opt = ptr.cast[Ptr[Byte]]
+    } else {
+      val ptr = stackalloc[CInt]
+      !ptr = value.asInstanceOf[Int]
+      opt = ptr.cast[Ptr[Byte]]
+    }
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: Could you refactor this one with a `match` too?

File: javalib/src/main/scala/java/net/Socket.scala [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    checkClosedAndCreate()
+    impl.setOption(SocketOptions.TCP_NODELAY, Boolean.box(on))
+  }
+
+  def setTrafficClass(tc: Int): Unit = {
+    checkClosedAndCreate()
+    impl.setOption(SocketOptions.IP_TOS, Integer.valueOf(tc))
+  }
+
+  def shutdownInput: Unit = {
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/Socket.scala]: `def shutdownInput()`

File: javalib/src/main/scala/java/net/Socket.scala [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+    impl.shutdownInput
+    inputShutdown = true
+  }
+
+  def shutdownOutput: Unit = {
```
Comment (Duhemm) [inline javalib/src/main/scala/java/net/Socket.scala]: `def shutdownOutput()`

File: javalib/src/main/scala/java/net/Socket.scala [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+  def this(host: String, port: Int, stream: Boolean) =
+    this(new PlainSocketImpl(),
+         InetAddress.getByName(host),
+         port,
+         null,
+         0,
+         true,
+         true)
+
+  // def this(proxy: Proxy)
```
Comment (asoltysik) [inline javalib/src/main/scala/java/net/Socket.scala]: #967

File: javalib/src/main/scala/java/net/PlainSocketImpl.scala [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+  protected[net] var localport            = 0
+  protected[net] var address: InetAddress = null
+  protected[net] var port                 = 0
+
+  private var timeout = 0
+
+  override def getInetAddress: InetAddress       = address
+  override def getFileDescriptor: FileDescriptor = fd
+
+  override def create(streaming: Boolean): Unit = {
```
Comment (asoltysik) [inline javalib/src/main/scala/java/net/PlainSocketImpl.scala]: It's a deprecated mode which makes `Socket` UDP instead of TCP. This can be delegated to `java.net.DatagramSocket` when it gets implemented.

https://github.com/asoltysik/scala-native/blob/1f24ccbefece8cbb4a21fe99ea6af6dbf9ee9ac1/javalib/src/main/scala/java/net/Socket.scala#L21-L25

File: javalib/src/main/scala/java/net/Socket.scala [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    startup(addr, port)
+  }
+
+  private def startup(dstAddr: InetAddress, dstPort: Int, timeout: Int = 0) = {
+    if (dstPort < 0 || dstPort > 65535)
+      throw new IllegalArgumentException(
+        "Socket port must be between 0 and 65535")
+
+    impl.create(streaming)
+    val created = true
```
Comment (asoltysik) [inline javalib/src/main/scala/java/net/Socket.scala]: I guess the socket descriptor was just created twice if the function `checkClosedAndCreate` was called :D Fixed

**Comment by asoltysik** [comment]
@Duhemm All comments addressed

**Comment by densh** [comment]
LGTM, fantastic work @asoltysik ! 👍
